File: ExamMonitor.java
import java.awt.*;
import java.awt.event.*;
import java.awt.image.BufferedImage;
import java.io.*;
import java.net.*;
import java.net.http.*;
import java.time.Duration;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class ExamMonitor {
    private static final String SERVER_URL = "http://127.0.0.1:8000";
    private static final int VIOLATION_LIMIT = 3;
    private static final int DEBOUNCE_MS = 2000;
    
    private int violationCount = 0;
    private long lastViolationTime = 0;
    private String currentAttemptId = null;
    private boolean isMonitoring = false;
    private ScheduledExecutorService scheduler;
    private TrayIcon trayIcon;
    private SystemTray systemTray;
    
    // Global keyboard hook for system-wide key detection
    private KeyboardHook keyboardHook;
    
    public ExamMonitor() {
        initializeTray();
        initializeKeyboardHook();
        startPeriodicCheck();
    }
    
    private void initializeTray() {
        if (SystemTray.isSupported()) {
            systemTray = SystemTray.getSystemTray();
            Image image = Toolkit.getDefaultToolkit().getImage(getClass().getResource("/icon.png"));
            if (image == null) {
                // Create a simple icon if resource not found
                image = createDefaultIcon();
            }
            
            PopupMenu popup = new PopupMenu();
            MenuItem statusItem = new MenuItem("Status: Waiting for exam");
            MenuItem exitItem = new MenuItem("Exit");
            
            exitItem.addActionListener(e -> {
                stopMonitoring();
                System.exit(0);
            });
            
            popup.add(statusItem);
            popup.addSeparator();
            popup.add(exitItem);
            
            trayIcon = new TrayIcon(image, "Exam Monitor", popup);
            trayIcon.setImageAutoSize(true);
            
            try {
                systemTray.add(trayIcon);
                trayIcon.displayMessage("Exam Monitor", "System monitoring active", TrayIcon.MessageType.INFO);
            } catch (AWTException e) {
                System.err.println("Failed to add tray icon: " + e.getMessage());
            }
        }
    }
    
    private Image createDefaultIcon() {
        BufferedImage image = new BufferedImage(16, 16, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g2d = image.createGraphics();
        g2d.setColor(Color.RED);
        g2d.fillOval(2, 2, 12, 12);
        g2d.dispose();
        return image;
    }
    
    private void initializeKeyboardHook() {
        try {
            // Use JNativeHook for global keyboard monitoring
            keyboardHook = new KeyboardHook();
            keyboardHook.start();
        } catch (Exception e) {
            System.err.println("Failed to initialize keyboard hook: " + e.getMessage());
            // Fallback to AWT key listener
            setupFallbackKeyboardMonitoring();
        }
    }
    
    private void setupFallbackKeyboardMonitoring() {
        // Create a hidden window to capture global key events
        Frame hiddenFrame = new Frame();
        hiddenFrame.setUndecorated(true);
        hiddenFrame.setSize(1, 1);
        hiddenFrame.setLocationRelativeTo(null);
        hiddenFrame.setVisible(false);
        
        hiddenFrame.addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                if (isMonitoring) {
                    checkForSuspiciousKeys(e);
                }
            }
        });
        
        hiddenFrame.addWindowFocusListener(new WindowAdapter() {
            @Override
            public void windowLostFocus(WindowEvent e) {
                if (isMonitoring) {
                    handleFocusLoss();
                }
            }
        });
    }
    
    private void checkForSuspiciousKeys(KeyEvent e) {
        // Detect Alt+Tab, Ctrl+Tab, F11, F12, etc.
        int keyCode = e.getKeyCode();
        boolean isAltPressed = e.isAltDown();
        boolean isCtrlPressed = e.isControlDown();
        
        if ((keyCode == KeyEvent.VK_TAB && (isAltPressed || isCtrlPressed)) ||
            keyCode == KeyEvent.VK_F11 || keyCode == KeyEvent.VK_F12 ||
            keyCode == KeyEvent.VK_ESCAPE) {
            
            handleViolation("Suspicious key combination detected: " + 
                          (isAltPressed ? "Alt+" : "") + 
                          (isCtrlPressed ? "Ctrl+" : "") + 
                          KeyEvent.getKeyText(keyCode));
        }
    }
    
    private void handleFocusLoss() {
        handleViolation("Application focus lost");
    }
    
    private void handleViolation(String reason) {
        long currentTime = System.currentTimeMillis();
        if (currentTime - lastViolationTime < DEBOUNCE_MS) {
            return; // Debounce rapid violations
        }
        
        lastViolationTime = currentTime;
        violationCount++;
        
        System.out.println("VIOLATION DETECTED: " + reason + " (Count: " + violationCount + "/" + VIOLATION_LIMIT + ")");
        
        if (trayIcon != null) {
            trayIcon.displayMessage("Exam Violation", 
                "Violation #" + violationCount + ": " + reason, 
                violationCount >= VIOLATION_LIMIT ? TrayIcon.MessageType.ERROR : TrayIcon.MessageType.WARNING);
        }
        
        // Send violation to web server
        sendViolationToServer(reason);
        
        if (violationCount >= VIOLATION_LIMIT) {
            handleExamTermination();
        }
    }
    
    private void sendViolationToServer(String reason) {
        if (currentAttemptId == null) return;
        
        try {
            HttpClient client = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(5))
                .build();
            
            String requestBody = "attempt_id=" + currentAttemptId;
            
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(SERVER_URL + "/increment_alt_tab"))
                .header("Content-Type", "application/x-www-form-urlencoded")
                .POST(HttpRequest.BodyPublishers.ofString(requestBody))
                .build();
            
            client.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenAccept(response -> {
                    if (response.statusCode() == 200) {
                        System.out.println("Violation sent to server successfully");
                    } else {
                        System.err.println("Failed to send violation to server: " + response.statusCode());
                    }
                })
                .exceptionally(e -> {
                    System.err.println("Error sending violation to server: " + e.getMessage());
                    return null;
                });
                
        } catch (Exception e) {
            System.err.println("Error sending violation to server: " + e.getMessage());
        }
    }
    
    private void handleExamTermination() {
        System.out.println("EXAM TERMINATED - Maximum violations reached!");
        
        if (trayIcon != null) {
            trayIcon.displayMessage("Exam Terminated", 
                "Exam terminated due to multiple violations!", 
                TrayIcon.MessageType.ERROR);
        }
        
        // Open results page in default browser
        try {
            Desktop.getDesktop().browse(new URI(SERVER_URL + "/results.html"));
        } catch (Exception e) {
            System.err.println("Failed to open results page: " + e.getMessage());
        }
        
        stopMonitoring();
    }
    
    private void startPeriodicCheck() {
        scheduler = Executors.newScheduledThreadPool(1);
        scheduler.scheduleAtFixedRate(() -> {
            if (isMonitoring) {
                checkForSuspiciousActivity();
            }
        }, 5, 5, TimeUnit.SECONDS);
    }
    
    private void checkForSuspiciousActivity() {
        // Check for minimized windows, inactive time, etc.
        try {
            // Check if any window is minimized
            GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
            GraphicsDevice[] devices = ge.getScreenDevices();
            
            for (GraphicsDevice device : devices) {
                GraphicsConfiguration[] configs = device.getConfigurations();
                for (GraphicsConfiguration config : configs) {
                    Rectangle bounds = config.getBounds();
                    if (bounds.width < 100 || bounds.height < 100) {
                        handleViolation("Window minimized or resized suspiciously");
                        return;
                    }
                }
            }
            
            // Check for extended inactivity (mouse/keyboard)
            // This would require additional system-level monitoring
            
        } catch (Exception e) {
            System.err.println("Error in periodic check: " + e.getMessage());
        }
    }
    
    public void startMonitoring(String attemptId) {
        this.currentAttemptId = attemptId;
        this.isMonitoring = true;
        this.violationCount = 0;
        
        System.out.println("Started monitoring for attempt ID: " + attemptId);
        
        if (trayIcon != null) {
            trayIcon.setToolTip("Exam Monitor - Active (Attempt: " + attemptId + ")");
            trayIcon.displayMessage("Exam Started", 
                "Monitoring active for attempt " + attemptId, 
                TrayIcon.MessageType.INFO);
        }
    }
    
    public void stopMonitoring() {
        this.isMonitoring = false;
        this.currentAttemptId = null;
        
        System.out.println("Stopped monitoring");
        
        if (trayIcon != null) {
            trayIcon.setToolTip("Exam Monitor - Inactive");
        }
    }
    
    public static void main(String[] args) {
        System.out.println("Starting Exam Monitor...");
        System.out.println("Server URL: " + SERVER_URL);
        
        ExamMonitor monitor = new ExamMonitor();
        
        // For testing, start monitoring with a dummy attempt ID
        if (args.length > 0) {
            monitor.startMonitoring(args[0]);
        }
        
        // Keep the application running
        try {
            Thread.currentThread().join();
        } catch (InterruptedException e) {
            System.out.println("Monitor interrupted");
        }
    }
}

// Global keyboard hook implementation
class KeyboardHook {
    private boolean running = false;
    
    public void start() {
        if (running) return;
        running = true;
        
        // This would use JNativeHook library for true global key monitoring
        // For now, we'll use the fallback AWT approach
        System.out.println("Keyboard hook started (fallback mode)");
    }
    
    public void stop() {
        running = false;
        System.out.println("Keyboard hook stopped");
    }
} 

File: exam_monitor.cpp
#include <windows.h>
#include <iostream>
#include <string>
#include <chrono>
#include <thread>
#include <curl/curl.h>
#include <json/json.h>

class ExamMonitor {
private:
    static ExamMonitor* instance;
    HHOOK keyboardHook;
    HHOOK mouseHook;
    int violationCount;
    long long lastViolationTime;
    std::string currentAttemptId;
    bool isMonitoring;
    std::string serverUrl;
    
    // Windows API callback functions
    static LRESULT CALLBACK KeyboardProc(int nCode, WPARAM wParam, LPARAM lParam);
    static LRESULT CALLBACK MouseProc(int nCode, WPARAM wParam, LPARAM lParam);
    static LRESULT CALLBACK WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam);
    
    // HTTP request helper
    static size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* userp);
    bool sendViolationToServer(const std::string& reason);
    
    // Violation handling
    void handleViolation(const std::string& reason);
    void handleExamTermination();
    
    // System monitoring
    void checkForSuspiciousActivity();
    void startPeriodicCheck();
    
public:
    ExamMonitor();
    ~ExamMonitor();
    
    void startMonitoring(const std::string& attemptId);
    void stopMonitoring();
    void run();
    
    // Getters
    bool getIsMonitoring() const { return isMonitoring; }
    int getViolationCount() const { return violationCount; }
};

// Static instance
ExamMonitor* ExamMonitor::instance = nullptr;

ExamMonitor::ExamMonitor() 
    : keyboardHook(nullptr), mouseHook(nullptr), violationCount(0), 
      lastViolationTime(0), isMonitoring(false), serverUrl("http://127.0.0.1:8000") {
    instance = this;
    
    // Initialize CURL
    curl_global_init(CURL_GLOBAL_ALL);
    
    std::cout << "Exam Monitor initialized" << std::endl;
    std::cout << "Server URL: " << serverUrl << std::endl;
}

ExamMonitor::~ExamMonitor() {
    if (keyboardHook) {
        UnhookWindowsHookEx(keyboardHook);
    }
    if (mouseHook) {
        UnhookWindowsHookEx(mouseHook);
    }
    curl_global_cleanup();
}

LRESULT CALLBACK ExamMonitor::KeyboardProc(int nCode, WPARAM wParam, LPARAM lParam) {
    if (nCode >= 0 && instance && instance->getIsMonitoring()) {
        if (wParam == WM_KEYDOWN) {
            KBDLLHOOKSTRUCT* kbStruct = (KBDLLHOOKSTRUCT*)lParam;
            
            // Check for Alt+Tab, Ctrl+Tab, F11, F12, etc.
            bool isAltPressed = GetAsyncKeyState(VK_MENU) & 0x8000;
            bool isCtrlPressed = GetAsyncKeyState(VK_CONTROL) & 0x8000;
            
            if ((kbStruct->vkCode == VK_TAB && (isAltPressed || isCtrlPressed)) ||
                kbStruct->vkCode == VK_F11 || kbStruct->vkCode == VK_F12 ||
                kbStruct->vkCode == VK_ESCAPE) {
                
                std::string reason = "Suspicious key combination detected: ";
                if (isAltPressed) reason += "Alt+";
                if (isCtrlPressed) reason += "Ctrl+";
                
                switch (kbStruct->vkCode) {
                    case VK_TAB: reason += "Tab"; break;
                    case VK_F11: reason += "F11"; break;
                    case VK_F12: reason += "F12"; break;
                    case VK_ESCAPE: reason += "Escape"; break;
                }
                
                instance->handleViolation(reason);
            }
        }
    }
    return CallNextHookEx(nullptr, nCode, wParam, lParam);
}

LRESULT CALLBACK ExamMonitor::MouseProc(int nCode, WPARAM wParam, LPARAM lParam) {
    if (nCode >= 0 && instance && instance->getIsMonitoring()) {
        // Monitor mouse activity for suspicious behavior
        // This could detect mouse leaving the exam window area
    }
    return CallNextHookEx(nullptr, nCode, wParam, lParam);
}

LRESULT CALLBACK ExamMonitor::WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam) {
    if (instance && instance->getIsMonitoring()) {
        switch (uMsg) {
            case WM_ACTIVATE:
                if (LOWORD(wParam) == WA_INACTIVE) {
                    instance->handleViolation("Window focus lost");
                }
                break;
            case WM_SIZE:
                if (wParam == SIZE_MINIMIZED) {
                    instance->handleViolation("Window minimized");
                }
                break;
        }
    }
    return DefWindowProc(hwnd, uMsg, wParam, lParam);
}

size_t ExamMonitor::WriteCallback(void* contents, size_t size, size_t nmemb, std::string* userp) {
    userp->append((char*)contents, size * nmemb);
    return size * nmemb;
}

bool ExamMonitor::sendViolationToServer(const std::string& reason) {
    if (currentAttemptId.empty()) return false;
    
    CURL* curl = curl_easy_init();
    if (!curl) return false;
    
    std::string postData = "attempt_id=" + currentAttemptId;
    std::string response;
    
    curl_easy_setopt(curl, CURLOPT_URL, (serverUrl + "/increment_alt_tab").c_str());
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, postData.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
    
    CURLcode res = curl_easy_perform(curl);
    long http_code = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
    
    curl_easy_cleanup(curl);
    
    if (res == CURLE_OK && http_code == 200) {
        std::cout << "Violation sent to server successfully" << std::endl;
        return true;
    } else {
        std::cerr << "Failed to send violation to server. HTTP code: " << http_code << std::endl;
        return false;
    }
}

void ExamMonitor::handleViolation(const std::string& reason) {
    auto now = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()).count();
    
    // Debounce rapid violations (2 seconds)
    if (now - lastViolationTime < 2000) {
        return;
    }
    
    lastViolationTime = now;
    violationCount++;
    
    std::cout << "VIOLATION DETECTED: " << reason 
              << " (Count: " << violationCount << "/3)" << std::endl;
    
    // Show system notification
    MessageBoxA(nullptr, 
        ("Violation #" + std::to_string(violationCount) + ": " + reason).c_str(),
        "Exam Violation", 
        violationCount >= 3 ? MB_ICONERROR : MB_ICONWARNING);
    
    // Send to server
    sendViolationToServer(reason);
    
    if (violationCount >= 3) {
        handleExamTermination();
    }
}

void ExamMonitor::handleExamTermination() {
    std::cout << "EXAM TERMINATED - Maximum violations reached!" << std::endl;
    
    MessageBoxA(nullptr, 
        "Exam terminated due to multiple violations!", 
        "Exam Terminated", 
        MB_ICONERROR);
    
    // Open results page in default browser
    ShellExecuteA(nullptr, "open", (serverUrl + "/results.html").c_str(), 
                  nullptr, nullptr, SW_SHOWNORMAL);
    
    stopMonitoring();
}

void ExamMonitor::checkForSuspiciousActivity() {
    // Check for minimized windows
    HWND foregroundWindow = GetForegroundWindow();
    if (foregroundWindow && IsIconic(foregroundWindow)) {
        handleViolation("Active window minimized");
    }
    
    // Check for extended inactivity (simplified)
    // In a real implementation, you'd track mouse/keyboard activity
}

void ExamMonitor::startPeriodicCheck() {
    std::thread([this]() {
        while (isMonitoring) {
            std::this_thread::sleep_for(std::chrono::seconds(5));
            if (isMonitoring) {
                checkForSuspiciousActivity();
            }
        }
    }).detach();
}

void ExamMonitor::startMonitoring(const std::string& attemptId) {
    currentAttemptId = attemptId;
    isMonitoring = true;
    violationCount = 0;
    
    std::cout << "Started monitoring for attempt ID: " << attemptId << std::endl;
    
    // Install hooks
    keyboardHook = SetWindowsHookEx(WH_KEYBOARD_LL, KeyboardProc, nullptr, 0);
    mouseHook = SetWindowsHookEx(WH_MOUSE_LL, MouseProc, nullptr, 0);
    
    if (!keyboardHook) {
        std::cerr << "Failed to install keyboard hook" << std::endl;
    }
    if (!mouseHook) {
        std::cerr << "Failed to install mouse hook" << std::endl;
    }
    
    startPeriodicCheck();
    
    MessageBoxA(nullptr, 
        ("Monitoring active for attempt " + attemptId).c_str(),
        "Exam Started", 
        MB_ICONINFORMATION);
}

void ExamMonitor::stopMonitoring() {
    isMonitoring = false;
    currentAttemptId.clear();
    
    if (keyboardHook) {
        UnhookWindowsHookEx(keyboardHook);
        keyboardHook = nullptr;
    }
    if (mouseHook) {
        UnhookWindowsHookEx(mouseHook);
        mouseHook = nullptr;
    }
    
    std::cout << "Stopped monitoring" << std::endl;
}

void ExamMonitor::run() {
    // Message loop
    MSG msg;
    while (GetMessage(&msg, nullptr, 0, 0)) {
        TranslateMessage(&msg);
        DispatchMessage(&msg);
    }
}

int main(int argc, char* argv[]) {
    std::cout << "Starting Exam Monitor (C++)..." << std::endl;
    
    ExamMonitor monitor;
    
    // For testing, start monitoring with a dummy attempt ID
    if (argc > 1) {
        monitor.startMonitoring(argv[1]);
    } else {
        monitor.startMonitoring("test_attempt_123");
    }
    
    // Run the message loop
    monitor.run();
    
    return 0;
} 